File: plugins/plugin-copilot-proxy/typescript/src/providers/modelProvider.ts
import {
  createOpenAICompatible,
  type OpenAICompatibleProvider,
} from "@ai-sdk/openai-compatible";
import type { IAgentRuntime } from "@elizaos/core";
import { logger } from "@elizaos/core";
import {
  getBaseUrl,
  getContextWindow,
  getLargeModel,
  getMaxTokens,
  getSmallModel,
  isBrowser,
  isPluginEnabled,
} from "../environment";
import type { ModelName, ValidatedBaseUrl } from "../types";

/**
 * Default models available through Copilot Proxy.
 */
export const DEFAULT_MODELS = [
  "gpt-5.2",
  "gpt-5.2-codex",
  "gpt-5.1",
  "gpt-5.1-codex",
  "gpt-5.1-codex-max",
  "gpt-5-mini",
  "claude-opus-4.5",
  "claude-sonnet-4.5",
  "claude-haiku-4.5",
  "gemini-3-pro",
  "gemini-3-flash",
  "grok-code-fast-1",
] as const;

export type DefaultModelId = (typeof DEFAULT_MODELS)[number];

/**
 * Model definition for Copilot Proxy models.
 */
export interface CopilotProxyModelDefinition {
  readonly id: string;
  readonly name: string;
  readonly api: "openai-completions";
  readonly reasoning: boolean;
  readonly input: readonly string[];
  readonly cost: {
    readonly input: number;
    readonly output: number;
    readonly cacheRead: number;
    readonly cacheWrite: number;
  };
  readonly contextWindow: number;
  readonly maxTokens: number;
}

/**
 * Build a model definition for a given model ID.
 */
export function buildModelDefinition(
  modelId: string,
  contextWindow: number = 128000,
  maxTokens: number = 8192,
): CopilotProxyModelDefinition {
  return {
    id: modelId,
    name: modelId,
    api: "openai-completions",
    reasoning: false,
    input: ["text", "image"],
    cost: { input: 0, output: 0, cacheRead: 0, cacheWrite: 0 },
    contextWindow,
    maxTokens,
  };
}

/**
 * Model provider configuration.
 */
export interface ModelProviderConfig {
  readonly baseUrl: ValidatedBaseUrl;
  readonly smallModel: ModelName;
  readonly largeModel: ModelName;
  readonly contextWindow: number;
  readonly maxTokens: number;
}

/**
 * Get model provider configuration from runtime.
 */
export function getModelProviderConfig(
  runtime: IAgentRuntime,
): ModelProviderConfig | null {
  if (!isPluginEnabled(runtime)) {
    logger.debug("[CopilotProxy] Plugin is disabled");
    return null;
  }

  const baseUrl = getBaseUrl(runtime);
  const smallModel = getSmallModel(runtime);
  const largeModel = getLargeModel(runtime);
  const contextWindow = getContextWindow(runtime);
  const maxTokens = getMaxTokens(runtime);

  return {
    baseUrl,
    smallModel,
    largeModel,
    contextWindow,
    maxTokens,
  };
}

/**
 * Cached provider instance.
 */
let cachedProvider: OpenAICompatibleProvider<string> | null = null;
let cachedBaseUrl: string | null = null;

/**
 * Create an OpenAI-compatible model provider for Copilot Proxy.
 *
 * This provider uses the Vercel AI SDK's OpenAI-compatible adapter
 * to communicate with the Copilot Proxy server.
 */
export function createCopilotProxyProvider(
  runtime: IAgentRuntime,
): OpenAICompatibleProvider<string> {
  const baseURL = isBrowser() ? undefined : getBaseUrl(runtime);
  const urlString = baseURL ?? "http://localhost:3000/v1";

  // Return cached provider if URL hasn't changed
  if (cachedProvider && cachedBaseUrl === urlString) {
    return cachedProvider;
  }

  const provider = createOpenAICompatible({
    name: "copilot-proxy",
    baseURL: urlString,
    // Copilot Proxy doesn't require authentication
    apiKey: "n/a",
    // Don't send authorization header
    headers: {},
  });

  cachedProvider = provider;
  cachedBaseUrl = urlString;

  return provider;
}

/**
 * Create a model instance for a specific model ID.
 */
export function createModelInstance(runtime: IAgentRuntime, modelId: string) {
  const provider = createCopilotProxyProvider(runtime);
  return provider(modelId);
}

/**
 * Get the small model instance.
 */
export function getSmallModelInstance(runtime: IAgentRuntime) {
  const modelId = getSmallModel(runtime);
  return createModelInstance(runtime, modelId);
}

/**
 * Get the large model instance.
 */
export function getLargeModelInstance(runtime: IAgentRuntime) {
  const modelId = getLargeModel(runtime);
  return createModelInstance(runtime, modelId);
}

/**
 * Check if a model ID is a known default model.
 */
export function isDefaultModel(modelId: string): modelId is DefaultModelId {
  return DEFAULT_MODELS.includes(modelId as DefaultModelId);
}

/**
 * Get all available model definitions.
 */
export function getAvailableModels(
  contextWindow: number = 128000,
  maxTokens: number = 8192,
): CopilotProxyModelDefinition[] {
  return DEFAULT_MODELS.map((modelId) =>
    buildModelDefinition(modelId, contextWindow, maxTokens),
  );
}

export { createCopilotProxyProvider as createProvider };

File: plugins/plugin-copilot-proxy/rust/src/providers/model_provider.rs
//! Model provider for Copilot Proxy.

use crate::config::{CopilotProxyConfig, AVAILABLE_MODELS};
use crate::service::CopilotProxyService;

/// Model definition for Copilot Proxy.
#[derive(Debug, Clone)]
pub struct ModelDefinition {
    /// The model ID.
    pub id: String,
    /// The model name.
    pub name: String,
    /// The API type.
    pub api: &'static str,
    /// Whether the model supports reasoning.
    pub reasoning: bool,
    /// Supported input types.
    pub input: Vec<&'static str>,
    /// Cost information.
    pub cost: ModelCost,
    /// Context window size.
    pub context_window: u32,
    /// Maximum tokens.
    pub max_tokens: u32,
}

/// Cost information for a model.
#[derive(Debug, Clone, Default)]
pub struct ModelCost {
    /// Input cost per token.
    pub input: f64,
    /// Output cost per token.
    pub output: f64,
    /// Cache read cost per token.
    pub cache_read: f64,
    /// Cache write cost per token.
    pub cache_write: f64,
}

impl ModelDefinition {
    /// Create a new model definition.
    pub fn new(
        id: impl Into<String>,
        context_window: u32,
        max_tokens: u32,
    ) -> Self {
        let id = id.into();
        Self {
            name: id.clone(),
            id,
            api: "openai-completions",
            reasoning: false,
            input: vec!["text", "image"],
            cost: ModelCost::default(),
            context_window,
            max_tokens,
        }
    }
}

/// Model provider configuration.
#[derive(Debug, Clone)]
pub struct ModelProviderConfig {
    /// Base URL for the proxy server.
    pub base_url: String,
    /// Small model ID.
    pub small_model: String,
    /// Large model ID.
    pub large_model: String,
    /// Context window size.
    pub context_window: u32,
    /// Maximum tokens.
    pub max_tokens: u32,
}

impl From<&CopilotProxyConfig> for ModelProviderConfig {
    fn from(config: &CopilotProxyConfig) -> Self {
        Self {
            base_url: config.base_url.clone(),
            small_model: config.small_model.clone(),
            large_model: config.large_model.clone(),
            context_window: config.context_window,
            max_tokens: config.max_tokens,
        }
    }
}

/// Get all available model definitions.
pub fn get_available_models(
    context_window: u32,
    max_tokens: u32,
) -> Vec<ModelDefinition> {
    AVAILABLE_MODELS
        .iter()
        .map(|id| ModelDefinition::new(*id, context_window, max_tokens))
        .collect()
}

/// Get the default available models.
pub fn get_default_models() -> Vec<ModelDefinition> {
    get_available_models(128_000, 8192)
}

/// Check if a model ID is a known model.
pub fn is_known_model(model_id: &str) -> bool {
    AVAILABLE_MODELS.contains(&model_id)
}

/// Model provider for Copilot Proxy.
pub struct CopilotProxyModelProvider {
    service: CopilotProxyService,
}

impl CopilotProxyModelProvider {
    /// Create a new model provider.
    pub fn new(config: CopilotProxyConfig) -> Self {
        Self {
            service: CopilotProxyService::new(config),
        }
    }

    /// Create a model provider from environment variables.
    pub fn from_env() -> Self {
        Self {
            service: CopilotProxyService::from_env(),
        }
    }

    /// Initialize the provider.
    pub async fn initialize(&self) -> crate::error::Result<()> {
        self.service.initialize().await
    }

    /// Check if the provider is available.
    pub async fn is_available(&self) -> bool {
        self.service.is_available().await
    }

    /// Get the service reference.
    pub fn service(&self) -> &CopilotProxyService {
        &self.service
    }

    /// Get the small model ID.
    pub fn small_model(&self) -> &str {
        self.service.small_model()
    }

    /// Get the large model ID.
    pub fn large_model(&self) -> &str {
        self.service.large_model()
    }

    /// Generate text using the small model.
    pub async fn generate_text_small(&self, prompt: &str) -> crate::error::Result<String> {
        self.service.generate_text_small(prompt).await
    }

    /// Generate text using the large model.
    pub async fn generate_text_large(&self, prompt: &str) -> crate::error::Result<String> {
        self.service.generate_text_large(prompt).await
    }

    /// Generate a JSON object using the small model.
    pub async fn generate_object_small(
        &self,
        prompt: &str,
    ) -> crate::error::Result<serde_json::Value> {
        self.service.generate_object_small(prompt).await
    }

    /// Generate a JSON object using the large model.
    pub async fn generate_object_large(
        &self,
        prompt: &str,
    ) -> crate::error::Result<serde_json::Value> {
        self.service.generate_object_large(prompt).await
    }

    /// Shutdown the provider.
    pub async fn shutdown(&self) {
        self.service.shutdown().await
    }
}
